Improve CMake

xsulca00 opened the pull request.

DDvO commented:
Hi, thank you for your PR.

Did you notice that there is another very related PR? See #42.

What would be the motivation/advantage of having a `CMakelists.txt` file in each source subdirectory?
Unfortunately I'm not a CMake expert, so I cannot see the pros and cons right away.
And why do you propose moving around all those source and header files?

DDvO commented:
Thanks a lot @xsulca00 for your explanations and sorry for my late response - I've been too busy.

Very nice to see that you have rebased this PR two days ago.
Meanwhile two new PRs of mine have been merged that further tweak `CMakeLists.txt`,
but I hope that the resulting new merge conflicts can be resolved without pain.

I'd love to learn more about good CMake practice, for instance from the book you mentioned,
but I fear I won't have the time for this in the near future.
Still I can imagine that it is helpful for having local `CMakeLists.txt` files in each directory.

> Moving those source and include files into `src` directory improves the project structure and maintainability, so one can have everything that belong together at one place.

I have not yet understood why moving 
* src/ → src/libsecutils/src/
* include/ → src/libsecutils/include/
* util/ → src/util/

is worth the major restructuring of all source files - can you please explain a little more?

@benjamin-schilling what do you think about this restructuring?

benjamin-schilling commented:
I agree that moving everything into the src/ directory creates a more cleaner repository structure.
In general if it makes integration with cmake easier I support it.

xsulca00 commented:
@benjamin-schilling explained it nicely. Thank you. It is also helpful if someone wants to search in the code of the specific component. It is enough just to pass the directory name.

I rebased everything and I think I also fixed the generated static_config.h file. Could you look at it please? Somehow the old static_config.h file persisted in the repo. I guess that is not what we want.

I would be thankful if we could merge this pull request as soon as possible because it takes a lot of time to rebase it regularly.

DDvO commented on src/libsecutils/include/secutils/credentials/verify.h:
```diff
@@ -71,6 +72,7 @@ int CREDENTIALS_print_cert_verify_cb(int ok, X509_STORE_CTX* ctx_x509);
 /* TODO DvO remove this function when the ones using it are merged upstream */
 bool verify_cb_cert(X509_STORE_CTX* store_ctx, X509* cert, int err);
 
+#ifdef SECUTILS_USE_UTA
```
No, this function needs to be available also without the UTA API !
Note that the `ura_ctx` is unused as documented. It is there just for backward compatibility.

DDvO commented on src/libsecutils/include/secutils/credentials/verify.h:
```diff
@@ -24,6 +24,7 @@
 #define X509_V_ERR_STORE_LOOKUP 70 /* from x509_vfy.h */
 #endif
 
+#include "../storage/uta_api.h"
```
Not needed, see below.

DDvO commented on src/util/CMakeLists.txt:
```diff
@@ -0,0 +1,54 @@
+    OUTPUT_NAME icvutil
+)
+
+target_compile_features(siemens_icvutil PRIVATE c_std_90)
+set_target_properties(siemens_icvutil PROPERTIES C_EXTENSIONS OFF)
+target_compile_options(siemens_icvutil PRIVATE
+    -Woverflow -Wswitch -Wmissing-prototypes -Wstrict-prototypes
+    -Wtype-limits -Wundef -Wconversion -Wno-shadow -Wno-conversion
+    -Wno-sign-conversion -Wno-unused-parameter -Wno-sign-compare # TODO enable -Wsign-conversion -Wsign-compare -Wunused-parameter
+    -Wformat -Wformat-security -Wno-declaration-after-statement -Wno-vla # -Wpointer-arith -pedantic -DPEDANTIC # -Werror
+    -pedantic -Werror
+)
+
```
It is not good to define this twcie.
Can't this be inherited here and in `src/libsecutils` by defining it once in `src/CMakeLists.txt`?

DDvO commented on CMakeLists.txt:
```diff
@@ -1,184 +1,21 @@
-project(secutils)
-include(GNUInstallDirs)
+project(
+  siemens_secutils
```
Moreover, better not use "siemens_" anywhere as part of file name etc.
Note that this is meanwhile general open-source software.
We should mention Siemens only in meta-texts such as copyright.

DDvO replied:
This one is also still open.

DDvO reviewed (changes requested):
I just had a closer look at all the changes.
I'm generally fine with the approach and agree it will improve maintainability.
Yet there are several incompatibilities, mistakes, and/or omissions to be sorted out.

DDvO commented on CMakeLists.txt:
```diff
@@ -1,184 +1,21 @@
-  ${INC_DIR}/util/log.h
-  ${INC_DIR}/operators.h
-  ${INC_DIR}/basic.h
-  ${INC_DIR}/static_config.h
```
My proposed solution is in #52. 
@xsulca00 please check and comment if this solves the issue you mentioned in https://github.com/siemens/libsecutils/pull/43#issuecomment-1558896514.

DDvO replied:
Please rebase on the meanwhile merged #52, which likely will solve this issue.

DDvO commented on CMakeLists.txt:
```diff
@@ -1,184 +1,21 @@
+option(SIEMENS_LIBSECUTILS_USE_ICV "Use configuration ICV" OFF)
+option(SIEMENS_LIBSECUTILS_NO_TLS "Do not use TLS" OFF)
 
-if(NOT("$ENV{OPENSSL_DIR}" STREQUAL ""))
-  message(STATUS "using OpenSSL from " $ENV{OPENSSL_DIR})
-  if(DEFINED ENV{OPENSSL_LIB})
-    set(OPENSSL_LIB $ENV{OPENSSL_LIB})
-  else()
-    set(OPENSSL_LIB $ENV{OPENSSL_DIR})
-  endif()
-  set(OpenSSL_INCLUDE_DIRS $ENV{OPENSSL_DIR}/include)
-  set(OPENSSL_LIBRARIES ${OPENSSL_LIB}/libssl.so ${OPENSSL_LIB}/libcrypto.so)
-  add_definitions(-isystem ${OpenSSL_INCLUDE_DIRS})
-else()
-  find_package(OpenSSL REQUIRED)
-  message(STATUS "using OpenSSL version " ${OPENSSL_VERSION})
-  add_definitions(-isystem /usr/include)
-endif()
```
@xsulca00 looks like this comment has not been handled yet.

DDvO commented on CMakeLists.txt:
```diff
@@ -1,184 +1,21 @@
 
-message(STATUS "SecurityUtilities version " ${SECUTILS_VERSION})
+find_package(OpenSSL MODULE REQUIRED COMPONENTS SSL)
 
```
This one is also still open.

DDvO commented on CMakeLists.txt:
```diff
@@ -1,184 +1,21 @@
-  set(CMAKE_INSTALL_PREFIX $ENV{ROOTFS} CACHE PATH "comment" FORCE)
-endif()
-
-install(FILES libsecutils.svg
-  DESTINATION ${CMAKE_INSTALL_FULL_DOCDIR}
-  )
-install(DIRECTORY doc/html/.
-  OPTIONAL  # TODO add build for doc/
-  DESTINATION ${CMAKE_INSTALL_FULL_DOCDIR}
-  )
```
Are you sure one can use `CMAKE_INSTALL_DOCDIR` instead of `CMAKE_INSTALL_FULL_DOCDIR` here?

DDvO commented on src/libsecutils/CMakeLists.txt:
```diff
@@ -0,0 +1,81 @@
+    -Woverflow -Wswitch -Wmissing-prototypes -Wstrict-prototypes
+    -Wtype-limits -Wundef -Wconversion -Wno-shadow -Wno-conversion
+    -Wno-sign-conversion -Wno-unused-parameter -Wno-sign-compare # TODO enable -Wsign-conversion -Wsign-compare -Wunused-parameter
+    -Wformat -Wformat-security -Wno-declaration-after-statement -Wno-vla # -Wpointer-arith -pedantic -DPEDANTIC # -Werror
```
Looks like this issue is gone meanwhile.

DDvO commented:
Hi @xsulca00, a couple of comments of my review on this PR are still open.
Can you handle or at least comment on them anytime soon?
I'd like to approve and merge this soon.

xsulca00 commented:
> Hi @xsulca00, a couple of comments of my review on this PR are still open. Can you handle or at least comment on them anytime soon? I'd like to approve and merge this soon.

Hi @DDvO , Im on vacation right now and Ive been busy in the last two weeks. On Monday, I will probably have time to look at it. Thanks!

DDvO commented:
@xsulca00 nice that you have rebased the PR again and have resumed handling the open review points last week...

xsulca00 commented:
> @xsulca00 when can you finish handling my review comments? I will be on vacation from Saturday for three weeks.

Sorry for misunderstanding, I though I answered all the comments and so I was waiting for your replies. Please just let me know if I am missing something. Thank you.

xsulca00 commented:
> @xsulca00 when can you finish handling my review comments? I will be on vacation from Saturday for three weeks.

So everything should be addressed now. Let me know please if I forgot something.

I use GitLab regularly which has a very different UI for dealing with review comments etc. so GitHub is somewhat confusing to me. Sorry for that.

DDvO reviewed (changes requested):
Unfortunately, as I feared, you did not catch and handle all open questions.
Please have a look at https://github.com/siemens/libsecutils/pull/43/files
where you should easily be able to notice them.

Moreover, I had to add several new questions and corrections.

Furthermore, the builds do not yet fully work with two of the three build options,
so please make sure that these work as well:
```
SECUTILS_USE_ICV= cmake . && make

SECUTILS_USE_UTA=1 cmake . && make
```

xsulca00 commented on src/libsecutils/CMakeLists.txt:
```diff
@@ -0,0 +1,76 @@
+)
+
+install(EXPORT siemens_libsecutilsTargets
+    DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/siemens
```
Fixed.

xsulca00 commented on src/libsecutils/CMakeLists.txt:
```diff
@@ -0,0 +1,76 @@
+install(FILES
+    siemens_libsecutilsConfig.cmake
+    ${CMAKE_CURRENT_BINARY_DIR}/siemens_libsecutilsConfigVersion.cmake
+    DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/siemens
```
Fixed.

xsulca00 reviewed (commented):
Sorry for this inconvenience, my mistake. I have started a "review" long time ago and in this mode GitHub accumulates comments. Sadly, I forgot to publish the review. So here we go.

xsulca00 commented on src/libsecutils/include/secutils/storage/uta_api.h:
```diff
@@ -16,6 +16,8 @@
 #ifndef SECUTILS_UTA_API_H_
 #define SECUTILS_UTA_API_H_
 
+#include "secutils_static_config.h"
+
```
Deleted, using common header basic.h.

xsulca00 commented on src/util/CMakeLists.txt:
```diff
@@ -0,0 +1,47 @@
+
+install(EXPORT siemens_icvutilTargets
+    DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/siemens
+    NAMESPACE siemens::
```
Installation directory for CMake configuration files changed to `cmake/security-utilities`.

xsulca00 commented on src/util/CMakeLists.txt:
```diff
@@ -0,0 +1,47 @@
+install(FILES
+    siemens_icvutilConfig.cmake
+    ${CMAKE_CURRENT_BINARY_DIR}/siemens_icvutilConfigVersion.cmake
+    DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/siemens
```
Installation directory for CMake configuration files changed to `cmake/security-utilities`.

xsulca00 commented on src/util/icvutil.c:
```diff
@@ -15,7 +15,7 @@
-#include "../include/secutils/util/log.h"
+#include "../libsecutils/include/secutils/util/log.h"
 #include "../include/secutils/storage/files_icv.h"
 #include "../include/secutils/storage/uta_api.h"
```
Done.

xsulca00 commented on src/util/icvutil.c:
```diff
@@ -66,8 +64,9 @@ int main(int argc, char *argv[]) {
-#ifdef SECUTILS_USE_UTA
     uta_close(uta_ctx);
+#else
+    LOG(FL_WARN, "Not using UTA lib because SECUTILS_USE_UTA was not defined");
```
Reverted.

DDvO commented on src/util/CMakeLists.txt:
```diff
@@ -0,0 +1,47 @@
+target_compile_features(security-utilities_icvutil PRIVATE c_std_90)
+set_target_properties(security-utilities_icvutil PROPERTIES C_EXTENSIONS ON)
+
+target_link_libraries(security-utilities_icvutil PRIVATE security-utilities::library secutils_compilation_options)
+
```
Here is missing:
```
    OpenSSL::Crypto
    secutils_compilation_options
```

Apparently you did not try
```
SECUTILS_USE_ICV=1 cmake . && make
```

DDvO replied:
I had to add: `$<$<BOOL:${SECURITY_UTILITIES_USE_UTA}>:security-utilities_uta-api>`

DDvO commented on CMakeLists.txt:
```diff
@@ -1,338 +1,69 @@
-set(CPACK_DEBIAN_DEV_PACKAGE_DEPENDS "libsecutils (>= ${CPACK_PACKAGE_VERSION})")
-set(CPACK_DEBIAN_DEV_PACKAGE_SUGGESTS "libssl-dev, libuta-dev")
-# see also https://gitlab.kitware.com/cmake/cmake/-/issues/21834
-# CPack Deb does not create postinst when installing to /usr/lib/x86_64-linux-gnu
-set(CPACK_DEBIAN_LIB_PACKAGE_CONTROL_EXTRA "${CMAKE_CURRENT_SOURCE_DIR}/debian/extra/triggers")
-set(CPACK_DEBIAN_LIB_PACKAGE_CONTROL_STRICT_PERMISSION TRUE)
-
-set(CPACK_SOURCE_GENERATOR "TGZ")
-#set(CPACK_SET_DESTDIR TRUE) # prevents package creation error when using cpack
-set(CPACK_VERBATIM_VARIABLES YES)
-set(CPACK_SOURCE_IGNORE_FILES
-  ${CPACK_IGNORE_FILES}
-  ~$
-  \./\.git/
-  \\.git$
-  \\.deb$
-  \\.gz$
-  \\.o$
-  \\.so
-  \./attic/
-  \./tmp/
-  CMakeFiles/
-  _CPack_Packages/
-  \\.cmake$
-  /CMakeCache.txt$
-  /compile_commands.json$
-  /install_manifest.*\.txt$
-  /changelog.gz$
-  \./debian/tmp/
-  \\.substvars$
-  \\.log$
-  /debian/\\.debhelper/
-  /debian/files$
-  /debian/debhelper-build-stamp$
-  \./debian/${PROJECT_NAME}/
-  \./debian/${PROJECT_NAME}-dev/
-  \./debian/${PROJECT_NAME}-bin/
-  \./doc/
-  \./icvutil$
-  )
```
All the Debian packaging stuff is gone - this needs to be retained.

DDvO commented on CMakeLists.txt:
```diff
@@ -1,338 +1,69 @@
-  COMPONENT lib
-  )
 
-find_program(DOXYGEN doxygen)
-if(NOT TARGET doc AND DOXYGEN)
-  add_custom_command(
-    DEPENDS "${CMAKE_CURRENT_SOURCE_DIR}/include/secutils/basic.h"
-    WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}"
-    COMMAND doxygen Doxyfile # required packages: doxygen graphviz
-    OUTPUT "${CMAKE_CURRENT_SOURCE_DIR}/doc/html" # BINARY_DIR woud be better
-    )
-  add_custom_target(doc ALL DEPENDS "${CMAKE_CURRENT_SOURCE_DIR}/doc/html")
-endif()
-
-install(DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/doc/html/."
-  OPTIONAL # as doxygen may not be available
-  DESTINATION "${CMAKE_INSTALL_FULL_DOCDIR}-dev/html"
-  COMPONENT dev
-  )
```
`make doc` does nothing - some definitions need to be re-added.

xsulca00 replied:
`make -f Makefile_v1 doc` runs fine to me. There is no main `Makefile` or you mean something else?

DDvO replied:
`make doc` should work as well, but it does nothing if the `doc` folder exists, and otherwise fails.

xsulca00 replied:
I see. You mean `Makefile` generated by CMake?

Doxygen added into CMake. It should work now.

DDvO replied:
Mostly. I had to add an `if(Doxygen_FOUND)`
and replace `set(DOXYGEN_OUTPUT_DIRECTORY doc/html)` by `set(DOXYGEN_OUTPUT_DIRECTORY doc)`

DDvO reviewed (changes requested):
Thanks for the recent improvements.
Yet some issues remain.

Also make sure that the following commands work again as expected (see `README.md`):
```
make doc
make deb
sudo make install
sudo make uninstall

xsulca00 commented on CMakeLists.txt:
```diff
@@ -1,338 +1,69 @@
 
+option(SIEMENS_LIBSECUTILS_USE_UTA "Use UTA API" OFF)
+option(SIEMENS_LIBSECUTILS_USE_ICV "Use configuration ICV" OFF)
+option(SIEMENS_LIBSECUTILS_NO_TLS "Do not use TLS" OFF)
```
They are not. Replaced with `SECURITY_UTILITIES`.

xsulca00 commented on src/CMakeLists.txt:
```diff
@@ -0,0 +1,21 @@
+target_compile_options(secutils_compilation_options INTERFACE
+    -Wall -Woverflow -Wextra -Wswitch -Wmissing-prototypes
+    -Wstrict-prototypes -Wformat -Wtype-limits -Wundef -Wconversion
+    -Wno-shadow -Wno-conversion -Wno-sign-conversion -Wno-sign-compare-Wno-unused-parameter
```
Thank you. Fixed.

xsulca00 commented:
Just as a side note, it is recommended to build CMake out-of-source build. Those command might be useful:
```
cmake -S . -B build_dir
cmake --build build_dir --parallel
cmake --install build_dir --prefix something (--component security-utilities_library_Runtime)
```

DDvO commented on src/libsecutils/CMakeLists.txt:
```diff
@@ -0,0 +1,73 @@
+    OpenSSL::SSL
+    OpenSSL::Crypto
+
+    $<$<BOOL:${SECURITY_UTILITIES_USE_UTA}>:-luta>
```
This does not yet work properly.
It works of you replace in by `$<$<BOOL:${SECUTILS_USE_UTA}>:-luta>`
and place earlier in this file:
```
if(SECURITY_UTILITIES_USE_UTA OR DEFINED ENV{SECUTILS_USE_UTA})
    set(SECUTILS_USE_UTA ON)
endif()
```

DDvO reviewed (changes requested):
Nice that `sudo make {un,}install` works now.
Yet `make doc` did not improve yet,
and `make deb` not yet (as you noted elsewhere).
Also `SECUTILS_USE_UTA=1 cmake . && make` needs a further tweak, see comment.

xsulca00 commented on CMakeLists.txt:
```diff
@@ -1,338 +1,69 @@
-  \./icvutil$
-  )
-
-include(CPack)
-
-# see also https://gitlab.kitware.com/cmake/cmake/-/issues/21832
-# CPack Deb lacks a way to configure copyright files by component
-foreach(component IN LISTS CPACK_COMPONENTS_ALL)
-  if(${component} STREQUAL "lib")
-    set(suffix "")
-  else()
-    set(suffix "-${component}")
-  endif()
-  install(FILES
-    "${CMAKE_CURRENT_SOURCE_DIR}/debian/copyright"
-    "${CMAKE_CURRENT_BINARY_DIR}/changelog.gz"
-    "${CMAKE_SOURCE_DIR}/libsecutils.svg"
-    DESTINATION "${CMAKE_INSTALL_FULL_DOCDIR}${suffix}"
-    COMPONENT ${component}
-    )
-endforeach()
-
-if(NOT TARGET deb)
-  set (deb_pack_full_name
-    "${CMAKE_CURRENT_BINARY_DIR}/${CPACK_PACKAGE_NAME}-dev_${CPACK_PACKAGE_VERSION}_all.deb")
-  add_custom_command(
-    OUTPUT "${deb_pack_full_name}"
-    DEPENDS build
-    COMMENT "Building packages possibly including doxygen run"
-    COMMAND cpack # --config CPackConfig.cmake
-    COMMAND cpack --config CPackSourceConfig.cmake
-    )
-  add_custom_target(deb DEPENDS ${deb_pack_full_name})
```
Did some changes, but I was busy and struggling with CPack for a while. Next week I have a vacation, so I will continue after that.

DDvO replied:
Good that you re-added that.
After some further tweaks, it seems to work all fine.

DDvO commented:
@xsulca00 do you have any news on the remaining items?

xsulca00 commented:
> @xsulca00 do you have any news on the remaining items?

CPack is still not ready yet. I forgot about it, Im trying to finish it now.

DDvO commented:
Sorry that I took that long to respond.
I was overly busy with other stuff, before I could take the time for consolidating my recent fixes to libsecutils and combining them with the latest version of your branch.

When I try `make deb` (after `cmake .`),
this fails because attempts are made to produce the following directories and write to them:
```
/usr/local/share/doc/security-utilities
/usr/local/share/doc/security-utilities-security-utilities_library_Development
/usr/local/share/doc/security-utilities-security-utilities_icvutil_Runtime
```
Something is wrong with the paths because any intermediate files should not be produced at `/usr/local/share/doc/` but locally.
As a workaround I created these three directories (with sudo) and made them writable for my user process.

Then `make deb` went through and produced some source archives, but no `.deb` files.

Moreover, the source archives contain too many files.
Compare for instance your `security-utilities-2.0.0-Source.tar.gz` with `libsecutils-2.0-Source.tar.gz` produced by `make deb` from the current master branch.

xsulca00 commented:
> Sorry that I took that long to respond. I was overly busy with other stuff, before I could take the time for consolidating my recent fixes to libsecutils and combining them with the latest version of your branch.
> 
> When I try `make deb` (after `cmake .`), this fails because attempts are made to produce the following directories and write to them:
> 
> ```
> /usr/local/share/doc/security-utilities
> /usr/local/share/doc/security-utilities-security-utilities_library_Development
> /usr/local/share/doc/security-utilities-security-utilities_icvutil_Runtime
> ```
> 
> Something is wrong with the paths because any intermediate files should not be produced at `/usr/local/share/doc/` but locally. As a workaround I created these three directories (with sudo) and made them writable for my user process.
> 
> Then `make deb` went through and produced some source archives, but no `.deb` files.
> 
> Moreover, the source archives contain too many files. Compare for instance your `security-utilities-2.0.0-Source.tar.gz` with `libsecutils-2.0-Source.tar.gz` produced by `make deb` from the current master branch.

Check please if the latests changes finally resolves all the points. I tried manually that:
* `cmake .` and `make deb` produce all packages and source package
* source package does not include extra files
* extra thing: one more package 'libsecutils-utils` is built when `SECURITY_UTILITIES_USE_ICV = ON` and contains `icvutil` binary

DDvO reviewed (approved):
LGTM, after I had to add a couple of further fixes and doc-related updates.

I also did some minor extensions, like a new `clean_all` target.

DDvO commented on src/libsecutils/CMakeLists.txt:
```diff
@@ -0,0 +1,81 @@
+    VERSION ${siemens_secutils_VERSION_MAJOR}.${siemens_secutils_VERSION_MINOR}.${siemens_secutils_VERSION_PATCH}
+)
+
+target_compile_features(siemens_libsecutils PRIVATE c_std_90)
```
I was not really resolved, as we recently discussed on Teams,
but meanwhile I've fixed it myself.

DDvO commented on src/libsecutils/src/storage/uta_api.c:
```diff
@@ -13,6 +13,8 @@
 
+#include <secutils/basic.h>
+
 #ifdef SECUTILS_USE_UTA 
 
 #include <openssl/sha.h>
```
The include path were fine then, but I had to add some non-trivial RPATH tweak found at
https://stackoverflow.com/questions/73248130/how-to-avoid-the-removal-of-the-rpath-during-cmake-install-step

``` 
if (APPLE)
  set(rbase "@loader_path")
else ()
  set(rbase "$ORIGIN")
endif ()
file(RELATIVE_PATH lib_dir
     "${CMAKE_CURRENT_BINARY_DIR}/${CMAKE_INSTALL_BINDIR}"
     "${CMAKE_CURRENT_BINARY_DIR}/${CMAKE_INSTALL_LIBDIR}")
# Honor user overrides, whether at the command line or FetchContent
set(CMAKE_INSTALL_RPATH "${rbase};${rbase}/${lib_dir}" CACHE STRING "Install RPATH")
```

DDvO commented:
@benjamin-schilling did you see this already?
Unfortunately, this PR involves many changes.
